numpy/_core/tests/test_stringdtype.py

```diff
@@ -0,0 +1,1374 @@
+    if use_out:
+        res = ufunc(arr, arr, out=arr)
+    else:
+        res = ufunc(arr, arr)
+
+    np.testing.assert_array_equal(uarr, res)
+
+
+@pytest.mark.parametrize("use_out", [True, False])
+@pytest.mark.parametrize(
+    "other_strings",
+    [
+        ["abc", "def" * 500, "ghi" * 16, "🤣" * 100, "📵", "😰"],
+        ["🚜", "🙃", "😾", "😹", "🚠", "🚌"],
+        ["🥦", "¨", "⨯", "∰ ", "⨌ ", "⎶ "],
+    ],
+)
+def test_ufunc_add(dtype, string_list, other_strings, use_out):
+    arr1 = np.array(string_list, dtype=dtype)
+    arr2 = np.array(other_strings, dtype=dtype)
+    result = np.array([a + b for a, b in zip(arr1, arr2)], dtype=dtype)
+
+    if use_out:
+        res = np.add(arr1, arr2, out=arr1)
+    else:
+        res = np.add(arr1, arr2)
+
+    np.testing.assert_array_equal(res, result)
+
+    if not hasattr(dtype, "na_object"):
+        return
+
+    is_nan = isinstance(dtype.na_object, float) and np.isnan(dtype.na_object)
+    is_str = isinstance(dtype.na_object, str)
+    bool_errors = 0
+    try:
+        bool(dtype.na_object)
+    except TypeError:
+        bool_errors = 1
+
+    arr1 = np.array([dtype.na_object] + string_list, dtype=dtype)
+    arr2 = np.array(other_strings + [dtype.na_object], dtype=dtype)
+
+    if is_nan or bool_errors or is_str:
+        res = np.add(arr1, arr2)
+        np.testing.assert_array_equal(res[1:-1], arr1[1:-1] + arr2[1:-1])
+        if not is_str:
+            assert res[0] is dtype.na_object and res[-1] is dtype.na_object
+        else:
+            assert res[0] == dtype.na_object + arr2[0]
+            assert res[-1] == arr1[-1] + dtype.na_object
+    else:
+        with pytest.raises(ValueError):
+            np.add(arr1, arr2)
+
+
+@pytest.mark.parametrize("use_out", [True, False])
+@pytest.mark.parametrize("other", [2, [2, 1, 3, 4, 1, 3]])
+@pytest.mark.parametrize(
+    "other_dtype",
+    [
+        None,
+        "int8",
+        "int16",
+        "int32",
+        "int64",
+        "uint8",
+        "uint16",
+        "uint32",
+        "uint64",
+        "short",
+        "int",
+        "intp",
+        "long",
+        "longlong",
+        "ushort",
+        "uint",
+        "uintp",
+        "ulong",
+        "ulonglong",
+    ],
+)
+def test_ufunc_multiply(dtype, string_list, other, other_dtype, use_out):
+    """Test the two-argument ufuncs match python builtin behavior."""
+    arr = np.array(string_list, dtype=dtype)
+    if other_dtype is not None:
+        other_dtype = np.dtype(other_dtype)
+    try:
+        len(other)
+        result = [s * o for s, o in zip(string_list, other)]
+        other = np.array(other)
+        if other_dtype is not None:
+            other = other.astype(other_dtype)
+    except TypeError:
+        if other_dtype is not None:
+            other = other_dtype.type(other)
+        result = [s * other for s in string_list]
+
+    if use_out:
+        arr_cache = arr.copy()
+        lres = np.multiply(arr, other, out=arr)
+        arr[:] = arr_cache
```

Comment:
good call, also added another `assert_array_equal` before `arr` and `lres` get reset.

Reply:
With the current code, I still do not understand how `lres` cannot have been reset to the original, and thus how the `assert_array_equal(lres, result)` can pass...  What am I missing? Is somehow `arr_cache` corrupted?

Reply:
it works because `rres` and `lres` are both `arr` at the end of the `if` statement because both calls to multiply use `out=arr`. Refactored to hopefully be less confusing.

Reply:
Ah, now I see it! I just missed the final time `arr` was overwritten...